// @lake/Lake.ts
import { assert } from './assert'
import { guard } from './guard'
import type { IComponent } from './types'
import { parseValue } from './html'
import { q } from './selector'

class Lake {
  definitions = new Map<{}, { fn: IComponent, deps: string[]}>()

  static create() {
    return new Lake()
  }

  define = (name: string, fn: IComponent, deps = [] as string[]) => {
    guard(this.definitions.has(name))
    this.definitions.set(name, { fn, deps })
  }

  resolve = (name: string) => {
    const context = this.definitions.get(name)
    assert(context, `${name} does not exist`)
    return context.fn
  }

  onSetup = () => {
    q(`[data-component]`).forEach(el => {
      const { component, props } = el.dataset
      guard(component)
      const context = this.resolve(component)
      context.setup(el, parseValue(props))
    })
  }

  onCleanup = (scope: HTMLElement) => {

  }
}

const __ = Lake.create()

const lake = __.definitions
const defineComponent = __.define
const resolveComponent = __.resolve

const onSetup = __.onSetup
const onCleanup = __.onCleanup

export { lake, defineComponent, resolveComponent, onSetup, onCleanup }
